Add support for dom.openOrClosedShadowRoot() and element.openOrClosedShadowRoot.

xeenon opened the pull request:
#### 109d3f887b571085b14b333091e7283dacc7f8b1
<pre>
Add support for dom.openOrClosedShadowRoot() and element.openOrClosedShadowRoot.
<a href="https://webkit.org/b/294340">https://webkit.org/b/294340</a>
<a href="https://rdar.apple.com/problem/153118095">rdar://problem/153118095</a>

Reviewed by Ryosuke Niwa.

Chrome supports `dom.openOrClosedShadowRoot()`, and Firefox supports `element.openOrClosedShadowRoot`.
Both APIs allow the same access to closed shadow roots from the extension content script world.

By adding a new `closedShadowRootIsExposedForExtensions` option on `DOMWrapperWorld` for extension content
script worlds, we can support both APIs with minimal additional work. We want to support both for broad
compatibility with existing extensions.

Added tests for both APIs.

This fixes the following WECG issue for Safari: <a href="https://github.com/w3c/webextensions/issues/815">https://github.com/w3c/webextensions/issues/815</a>

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::canAccessAnyShadowRoot const): Added.
(WebCore::DOMWrapperWorld::setClosedShadowRootIsExposedForExtensions): Added.
(WebCore::DOMWrapperWorld::closedShadowRootIsExposedForExtensions const): Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::openOrClosedShadowRoot const): Added.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::composedPath const): Use canAccessAnyShadowRoot().
* Source/WebKit/CMakeLists.txt: Added WebExtensionAPIDOM.
* Source/WebKit/DerivedSources-input.xcfilelist: Updated.
* Source/WebKit/DerivedSources-output.xcfilelist: Updated.
* Source/WebKit/DerivedSources.make: Added WebExtensionAPIDOM.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDOMCocoa.mm: Copied from Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.h.
(WebKit::WebExtensionAPIDOM::openOrClosedShadowRoot):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::dom):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDOM.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIDOM.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::setContentScriptWorld):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInScriptWorld.mm:
(-[WKWebProcessPlugInScriptWorld exposeClosedShadowRootsForExtensions]): Added.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.cpp:
(WKBundleScriptWorldExposeClosedShadowRootsForExtensions): Added.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.h:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::exposeClosedShadowRootsForExtensions): Added.
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt: Added WKWebExtensionAPIDOM.mm.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDOM.mm: Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIDOM, OpenOrClosedShadowRoot)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIDOM, OpenOrClosedShadowRootViaElement)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm: Added missing include.

Canonical link: <a href="https://commits.webkit.org/296168@main">https://commits.webkit.org/296168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb1cc53aa8bc078c2859bb06558ab2f84af51c41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107572 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17665 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/112789 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109536 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27946 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35757 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/112789 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110501 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/96956 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/112789 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/57553 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/15129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/115890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34640 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/115890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35016 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/93206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/115890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/35366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17398 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/34561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/34307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37667 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/35970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
```diff
@@ -186,6 +186,12 @@ void WebExtensionContextProxy::setStorageAccessLevel(bool allowedInContentScript
+void WebExtensionContextProxy::setContentScriptWorld(WebCore::DOMWrapperWorld& world)
+{
+    m_contentScriptWorld = world;
+    world.setShadowRootIsAlwaysOpen();
```
I don't think we want to set this option if we're adding a new API to access closed shadow roots.
This API also changes the behavior of event.composedPath, etc...

xeenon replied:
Extensions have asked for access to closed roots in `event.composedPath` too. We want to enable this.

rniwa replied:
Do other browser really expose nodes from closed shadow trees in `event.composedPath`? I don't think we want to diverge from other browsers here.

xeenon replied:
Chrome and Firefox have said that are planning to do it in when it came up in the WECG.

rniwa replied:
Okay, in that case, we should add a new DOM wrapper option for extensions and expose nodes in closed shadow trees when the option is set in addition to exposing new API to extensions.

But we definitely don't want to expose `collectMatchingElementsInFlatTree` or `matchingElementInFlatTree`.

rniwa replied:
`shadowRootIsAlwaysOpen` also changes the behavior of `Element.prototype.shadowRoot` and I don't think we want to do that either.

xeenon replied:
Alright. I'll revise then. Thanks!

xeenon replied:
@rniwa Updated this PR.

rniwa commented on Source/WebCore/dom/Element.idl:
```diff
@@ -60,6 +60,7 @@
 
     ShadowRoot attachShadow(ShadowRootInit init);
     [ImplementedAs=shadowRootForBindings, CallWith=CurrentGlobalObject] readonly attribute ShadowRoot? shadowRoot;
+    [EnabledForWorld=shadowRootIsAlwaysOpen] readonly attribute ShadowRoot? openOrClosedShadowRoot;
```
I don't think we want to expose this to every world which forces shadow root be always open.
e.g. DOM wrapper world Safari uses to implement autofill has no business having access to this API.

rniwa reviewed (changes requested):
I don't think we want expose element.openOrClosedShadowRoot to the DOM wrapper world Safari uses to implement autofill, and vice versa. We most definitely don't want to expose `collectMatchingElementsInFlatTree` or `matchingElementInFlatTree` to extensions.

rniwa reviewed (approved):
Looks good to me!

xeenon committed:
Add support for dom.openOrClosedShadowRoot() and element.openOrClosedShadowRoot.
https://webkit.org/b/294340
rdar://problem/153118095

Reviewed by Ryosuke Niwa.

Chrome supports `dom.openOrClosedShadowRoot()`, and Firefox supports `element.openOrClosedShadowRoot`.
Both APIs allow the same access to closed shadow roots from the extension content script world.

By adding a new `closedShadowRootIsExposedForExtensions` option on `DOMWrapperWorld` for extension content
script worlds, we can support both APIs with minimal additional work. We want to support both for broad
compatibility with existing extensions.

Added tests for both APIs.

This fixes the following WECG issue for Safari: https://github.com/w3c/webextensions/issues/815

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::canAccessAnyShadowRoot const): Added.
(WebCore::DOMWrapperWorld::setClosedShadowRootIsExposedForExtensions): Added.
(WebCore::DOMWrapperWorld::closedShadowRootIsExposedForExtensions const): Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::openOrClosedShadowRoot const): Added.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/dom/Event.cpp:
(WebCore::Event::composedPath const): Use canAccessAnyShadowRoot().
* Source/WebKit/CMakeLists.txt: Added WebExtensionAPIDOM.
* Source/WebKit/DerivedSources-input.xcfilelist: Updated.
* Source/WebKit/DerivedSources-output.xcfilelist: Updated.
* Source/WebKit/DerivedSources.make: Added WebExtensionAPIDOM.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDOMCocoa.mm: Copied from Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.h.
(WebKit::WebExtensionAPIDOM::openOrClosedShadowRoot):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::dom):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDOM.h: Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::globalObjectIsAvailableForFrame):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIDOM.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::setContentScriptWorld):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInScriptWorld.mm:
(-[WKWebProcessPlugInScriptWorld exposeClosedShadowRootsForExtensions]): Added.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.cpp:
(WKBundleScriptWorldExposeClosedShadowRootsForExtensions): Added.
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundleScriptWorld.h:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::exposeClosedShadowRootsForExtensions): Added.
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Tools/TestWebKitAPI/SourcesCocoa.txt: Added WKWebExtensionAPIDOM.mm.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDOM.mm: Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIDOM, OpenOrClosedShadowRoot)): Added.
(TestWebKitAPI::TEST(WKWebExtensionAPIDOM, OpenOrClosedShadowRootViaElement)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITest.mm: Added missing include.

Canonical link: https://commits.webkit.org/296168@main
